File: messagebus/src/main/java/com/yahoo/messagebus/network/rpc/RPCServiceAddress.java
// Copyright Vespa.ai. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.messagebus.network.rpc;

import com.yahoo.messagebus.network.ServiceAddress;
import com.yahoo.jrt.Spec;

/**
 * Implements the {@link ServiceAddress} interface for the RPC network.
 *
 * @author <a href="mailto:havardpe@yahoo-inc.com">Haavard Pettersen</a>
 */
public class RPCServiceAddress implements ServiceAddress {

    private final String serviceName;
    private final String sessionName;
    private final Spec connectionSpec;
    private RPCTarget target;

    /**
     * Constructs a service address from the given specifications. The last component of the service is stored as the
     * session name.
     *
     * @param serviceName    The full service name of the address.
     * @param connectionSpec The connection specification.
     */
    public RPCServiceAddress(String serviceName, Spec connectionSpec) {
        this.serviceName = serviceName;
        int pos = serviceName.lastIndexOf('/');
        sessionName =  (pos > 0 && pos < serviceName.length() - 1)
                ? serviceName.substring(pos + 1)
                : null;
        this.connectionSpec = connectionSpec;
    }
    public RPCServiceAddress(String serviceName, String connectionSpec) {
        this(serviceName, new Spec(connectionSpec));
    }
    public RPCServiceAddress(RPCServiceAddress blueprint) {
        serviceName = blueprint.serviceName;
        sessionName = blueprint.sessionName;
        connectionSpec = blueprint.connectionSpec;
        target = null;
    }


    @Override
    public boolean equals(Object obj) {
        if (obj == null || !(obj instanceof RPCServiceAddress)) {
            return false;
        }
        RPCServiceAddress rhs = (RPCServiceAddress)obj;
        if (!serviceName.equals(rhs.serviceName)) {
            return false;
        }
        if (!connectionSpec.host().equals(rhs.connectionSpec.host())) {
            return false;
        }
        if (connectionSpec.port() != rhs.connectionSpec.port()) {
            return false;
        }
        if (!sessionName.equals(rhs.sessionName)) {
            return false;
        }
        return true;
    }

    @Override
    public int hashCode() {
        return serviceName.hashCode() + connectionSpec.hashCode() + sessionName.hashCode();
    }

    /**
     * Returns whether or not this service address is malformed.
     *
     * @return True if malformed.
     */
    public boolean isMalformed() {
        return sessionName == null || connectionSpec.malformed();
    }

    /**
     * Returns the name of the remove service.
     *
     * @return The service name.
     */
    public String getServiceName() {
        return serviceName;
    }

    /**
     * Returns the name of the remote session.
     *
     * @return The session name.
     */
    public String getSessionName() {
        return sessionName;
    }

    /**
     * Returns the connection spec for the remote service.
     *
     * @return The connection spec.
     */
    public Spec getConnectionSpec() {
        return connectionSpec;
    }

    /**
     * Sets the RPC target to be used when communicating with the remote service.
     *
     * @param target The target to set.
     */
    public void setTarget(RPCTarget target) {
        this.target = target;
    }

    /**
     * Returns the RPC target to be used when communicating with the remove service.
     *
     * @return The target to use.
     */
    public RPCTarget getTarget() {
        return target;
    }
}

File: messagebus/src/vespa/messagebus/network/rpcserviceaddress.cpp
// Copyright Vespa.ai. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
#include "rpcserviceaddress.h"

namespace mbus {

RPCServiceAddress::RPCServiceAddress(const string &serviceName,
                                     const string &connectionSpec) :
    _serviceName(serviceName),
    _sessionName(""),
    _connectionSpec(connectionSpec),
    _target()
{
    size_t pos = serviceName.find_last_of('/');
    if (pos != string::npos) {
        _sessionName = serviceName.substr(pos + 1);
    }
}

RPCServiceAddress::~RPCServiceAddress() = default;

bool
RPCServiceAddress::isMalformed()
{
    if (_serviceName.empty()) {
        return true; // no service
    }
    if (_sessionName.empty()) {
        return true; // no session
    }
    if (_connectionSpec.empty()) {
        return true; // no spec
    }
    size_t prefixPos = _connectionSpec.find("tcp/");
    if (prefixPos != 0) {
        return true; // no prefix
    }
    size_t portPos = _connectionSpec.find_first_of(':', prefixPos);
    if (portPos == string::npos) {
        return true; // not colon
    }
    if (portPos == prefixPos + 4) {
        return true; // no address
    }
    if (portPos == _connectionSpec.size() - 1) {
        return true; // no port
    }
    return false;
}

} // namespace mbus
